### PR TITLE
time measurements between tracepoints

### PR DESCRIPTION
**Pull request by lievenhey**
this patch allows the user to measure the time it takes between two
tracepoints and shows this as a cost

### CODE REVIEW

**Comment by milianw** [comment]
@lievenhey how did you get the broken perf data file? And what exactly is broken - that you get two consecutive enter events instead of enter/exit pairs?

File: src/parsers/perf/perfparser.cpp [hunk 1]
```diff
@@ -1009,6 +1012,38 @@ class PerfParserPrivate : public QObject
         }
     }
 
+    void addTracepointToTracepointStack(uint32_t tid, const Data::Tracepoint& tracepoint)
+    {
+        const QVector<Data::TracepointTimeMeasurementsParameters> tracepoints =
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1017]: const auto

File: src/parsers/perf/perfparser.cpp [hunk 2]
```diff
@@ -1009,6 +1012,38 @@ class PerfParserPrivate : public QObject
...
+            Settings::instance()->tracepointParameters();
+
+        auto addToLookup = [this, &tracepoints](const QString& name) -> bool {
+            for (const auto& parameters : tracepoints) {
+                auto match = parameters.startRegex.match(name);
+                if (match.hasMatch()) {
+                    QString stopName = parameters.stopRegex.pattern();
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1024]: huh what is this doing here? why do you take the pattern and make it a string? this looks very odd to me - don't you instead just want to store the stop regex directly and use that later on? or match the stopregex directly and replace some of its matches somehow? but taking the pattern and doing stuff with its string sounds totally backwards - why is it a regex then to begin with?

File: src/parsers/perf/perfparser.cpp [hunk 3]
```diff
@@ -1035,12 +1070,51 @@ class PerfParserPrivate : public QObject
...
+                    if (tracepoint.name == exspectedTracepointName[sample.tid].last()) {
+                        const auto startTracepoint = tracepointStack[sample.tid].back();
+
+                        bottomUpResult.addTracepointCost(tracepointCostIdLookup[startTracepoint.name], sample.frames,
+                                                         tracepoint.time - startTracepoint.time);
+
+                        tracepointStack[sample.tid].pop_back();
+                        exspectedTracepointName[sample.tid].pop_back();
+                    } else {
+                        // TODO: find solution for this
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1093]: as I mentioned, how did you get this?

but yes, e.g. in the face of lost events (you'll see an error/warning during perf record) this can happen. or if we only trace enter and not the exit events. or if we have user tracepoints and those don't get emitted in the correct order. 

so one way or another this needs to be handled gracefully. I'd just say emit a warning or such and override the previous start entry

File: src/settingsdialog.cpp [hunk 4]
```diff
@@ -26,9 +27,9 @@
 #include "hotspot-config.h"
 
 namespace {
-KConfigGroup config()
+auto config()
```
Comment (milianw) [inline src/settingsdialog.cpp:30]: use KSharedConfig::Ptr here please instead of auto, I don't think we rely on c++17 yet? or is this valid c++14 already?

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/parsers/perf/perfparser.cpp (hunk 3 above)
Comment (lievenhey) [reply src/parsers/perf/perfparser.cpp:1093]: I used `perf record -o /home/lieven/perf.data --call-graph dwarf,8192 --event cycles:u,syscalls:sys_enter_openat,syscalls:sys_exit_openat --aio --sample-cpu /home/lieven/KDAB/hotspot/build/tests/test-clients/c-syscalls/c-syscalls`.
Hotspot reports lost events.

File: src/settingsdialog.cpp (hunk 4 above)
Comment (lievenhey) [reply src/settingsdialog.cpp:30]: seems to be valid, but I changed it anyway

File: src/parsers/perf/perfparser.cpp (hunk 2 above)
Comment (lievenhey) [reply src/parsers/perf/perfparser.cpp:1024]: I tried implementing regex support as requested in #346. To improve performance I wanted to cache the regexes. It seems  I didn't notice that parameters.stopRegex doesn't need to be a regex

File: src/parsers/perf/perfparser.cpp (hunk 3 above)
Comment (milianw) [reply src/parsers/perf/perfparser.cpp:1093]: right, that's somewhat expected then. try to limit the syscalls (e.g. only look at openat or smiilar). or enable compression via `-z` to resolve the lost chunks.

but what you are seeing here is a "reality" we need to live with in the parser frontend

File: src/models/data.h [hunk 5]
```diff
@@ -804,6 +807,13 @@ struct ZoomAction
         return time.isValid();
     }
 };
+
+struct TracepointTimeMeasurementsParameters
+{
+    QRegularExpression startRegex;
+    QString stopRegex;
```
Comment (milianw) [inline src/models/data.h:814]: this isn't a regex, it's actually a kind of replacement that is fed into the start regex. can you please add a comment that explains the relation and use better names? because clearly stopRegex isn't a regex, or?

File: src/parsers/perf/perfparser.cpp [hunk 6]
```diff
@@ -812,6 +812,9 @@ class PerfParserPrivate : public QObject
     {
         Data::BottomUp::initializeParents(&bottomUpResult.root);
 
+        for (auto it = tracepointCostIdLookup.begin(); it != tracepointCostIdLookup.end(); it++) {
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:815]: can you add a comment why this is needed? why do we want to have the total application time as total cost? is this just for tooltips to compute a sane relative cost?

File: src/parsers/perf/perfparser.cpp [hunk 7]
```diff
@@ -1009,6 +1012,37 @@ class PerfParserPrivate : public QObject
         }
     }
 
+    void addTracepointToTracepointStack(uint32_t tid, const Data::Tracepoint& tracepoint)
+    {
+        const auto tracepoints = Settings::instance()->tracepointParameters();
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1017]: this is not safe to be called from a background thread

generally, please add (in a separate patch) a `Q_ASSERT(QThread::currentThread() == settings.thread())` (and then compile with assertions enabled) to catch this in the future :)

File: src/parsers/perf/perfparser.cpp [hunk 8]
```diff
@@ -1465,6 +1528,9 @@ void PerfParser::startParseFile(const QString& path, const QString& sysroot, con
                 emit parserWarning(tr("Samples contained no call stack frames. Consider passing <code>--call-graph "
                                       "dwarf</code> to <code>perf record</code>."));
             }
+            if (d.droppedTracepoints) {
+                emit parserWarning(tr("Encountered unexspected tracepoints. Derived cost may be inaccurate."));
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1532]: typo: unexpected

File: src/resultspage.cpp [hunk 9]
```diff
@@ -103,7 +103,6 @@ ResultsPage::ResultsPage(PerfParser* parser, QWidget* parent)
         dockify(m_resultsDisassemblyPage, QStringLiteral("disassembly"), tr("D&isassembly"), tr("Ctrl+I"));
     m_summaryPageDock->addDockWidgetAsTab(m_disassemblyDock, KDDockWidgets::InitialVisibilityOption::StartHidden);
     m_disassemblyDock->toggleAction()->setEnabled(false);
-    m_summaryPageDock->setAsCurrentTab();
```
Comment (milianw) [inline src/resultspage.cpp:106]: why this change? seems unrelated?

File: src/settings.h [hunk 10]
```diff
@@ -10,6 +10,9 @@
 
 #include <QColor>
 #include <QObject>
+#include <QVector>
+
+#include "models/data.h"
```
Comment (milianw) [inline src/settings.h:15]: forward declare the Parameters struct instead please

File: src/settingsdialog.cpp [hunk 11]
```diff
@@ -290,3 +293,37 @@ void SettingsDialog::addCallgraphPage()
...
+
+    auto config = new MultiConfigWidget();
+
+    connect(config, &MultiConfigWidget::saveConfig, this, saveFunction);
+    connect(config, &MultiConfigWidget::restoreConfig, this, restoreFunction);
+
+    config->setConfig(::config()->group("TracepointsTimeMeasurements"));
+    config->restoreCurrent();
+
+    tracepointPage->layout->insertRow(0, QLatin1String("Name:"), config);
```
Comment (milianw) [inline src/settingsdialog.cpp:326]: this QL1S should be a tr()

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/settings.h (hunk 10 above)
Comment (lievenhey) [reply src/settings.h:15]: not possible, but I moved the struct to settings.h

File: src/parsers/perf/perfparser.cpp (hunk 6 above)
Comment (lievenhey) [reply src/parsers/perf/perfparser.cpp:815]: I wasn't sure what to use as total cost. I thought that if hotspot tells you, that function x spends 30% of its runtime in syscall y, its quite useful.

Comment (lievenhey) [reply src/parsers/perf/perfparser.cpp:815]: But the other way around (30% of the runtime in syscall x was spent in function y) is also valid.
Maybe I should add a settings to choose ether application runtime or total syscall runtime as totalCost.

Comment (milianw) [reply src/parsers/perf/perfparser.cpp:815]: ok, then keep it as-is and add a comment to indicate that this is to get costs relative to the total runtime in tooltips etc.

File: src/parsers/perf/perfparser.cpp [hunk 12]
```diff
@@ -1009,6 +1012,35 @@ class PerfParserPrivate : public QObject
         }
     }
 
+    void addTracepointToTracepointStack(uint32_t tid, const Data::Tracepoint& tracepoint)
+    {
+        auto addToLookup = [this](const QString& name) -> bool {
+            for (const auto& parameters : tracepointParameters) {
+                auto match = parameters.startRegex.match(name);
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1019]: ok, now that I understand what you are doing here:

isn't this the same as

```
auto stopName = name.replace(parameters.startRegex, parameters.stopExpression);
if (stopname == name)
    continue; // not matched
```

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/parsers/perf/perfparser.cpp (hunk 12 above)
Comment (lievenhey) [reply src/parsers/perf/perfparser.cpp:1024]: yes

File: src/models/treemodel.cpp [hunk 13]
```diff
@@ -54,6 +54,10 @@ QVariant BottomUpModel::headerColumnData(int column, int role) const
                 "The name of the executable the symbol resides in. May be empty when debug information is missing.");
         }
 
+        if (m_results.costs.unit(column - NUM_BASE_COLUMNS) == Data::Costs::Unit::TracepointCost) {
+            return tr("The total time spend in the function.");
```
Comment (milianw) [inline src/models/treemodel.cpp:58]: in which function? this should be rephrased to include the typeName (which is the tracepoint)

File: src/models/treemodel.cpp [hunk 14]
```diff
@@ -136,6 +140,10 @@ QVariant TopDownModel::headerColumnData(int column, int role) const
         }
         column -= NUM_BASE_COLUMNS;
         if (column < m_results.inclusiveCosts.numTypes()) {
+            if (m_results.inclusiveCosts.unit(column) == Data::Costs::Unit::TracepointCost) {
+                return tr("The total time spend in the function.");
+            }
+
```
Comment (milianw) [inline src/models/treemodel.cpp:146]: dito

File: src/parsers/perf/perfparser.cpp [hunk 15]
```diff
@@ -1040,12 +1071,39 @@ class PerfParserPrivate : public QObject
...
                 tracepoint.name = strings.value(attribute.name.id);
                 if (tracepoint.name != QLatin1String("sched:sched_switch")) {
                     // sched_switch events are handled separately already
                     tracepointResult.tracepoints.push_back(tracepoint);
                 }
+
+                if (exspectedTracepointName[sample.tid].isEmpty()) {
+                    addTracepointToTracepointStack(sample.tid, tracepoint);
+                } else {
+                    if (tracepoint.name == exspectedTracepointName[sample.tid].last()) {
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1084]: typo expected

File: src/parsers/perf/perfparser.cpp [hunk 16]
```diff
@@ -1755,6 +1845,17 @@ void PerfParser::filterResults(const Data::FilterAction& filter)
...
             Data::BottomUp::initializeParents(&bottomUp.root);
 
+            QSet<qint32> costIds;
+            for (const auto& cost : tracepointCost) {
+                costIds.insert(cost.costId);
+                bottomUp.addTracepointCost(cost);
+            }
+
+            for (const auto costIds : costIds) {
+                qDebug() << filter.time.delta();
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1855]: remove

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/models/treemodel.cpp (hunk 13 above)
Comment (lievenhey) [reply src/models/treemodel.cpp]: `tr("The total time spend in %1.").arg(m_results.costs.typeName(column - NUM_BASE_COLUMNS))`

File: src/parsers/perf/perfparser.cpp [hunk 17]
```diff
@@ -1015,6 +1018,34 @@ class PerfParserPrivate : public QObject
...
+                }
+
+                tracepointNameRegexCache[name] = stopName;
+                tracepointCostIdLookup[name] = addCostType(parameters.name, Data::Costs::Unit::TracepointCost);
+                return true;
+            }
+            return false;
+        };
+
+        if (!tracepointNameRegexCache.contains(tracepoint.name)) {
```
Comment (milianw) [inline src/parsers/perf/perfparser.cpp:1039]: multiple double qhash lookups (once for contains, once for write/lookup) and we should cache negative lookups too. more duplicate lookups in the single place this is being used from (addSample), so rewrite this all as:

```
QString toTracepointStopName(const QString &name)
{
auto cacheIt = tracepointStopNames.find(name);
if (cacheIt == tracepointStopNames.end()) {
    for (const auto& parameters : tracepointParameters) {
        auto stopName = name;
        stopName.replace(parameters.startRegex, parameters.stopExpression);

        if (stopName == name) {
            continue;
        }

        cacheIt = tracepointStopNames.insert(stopName);
        tracepointCostIdLookup[name] = addCostType(parameters.name, Data::Costs::Unit::TracepointCost);
        break;
    }
    if (cacheIt == tracepointStopNames.end()) {
        // not a tracepoint that should be time-measured
        cacheIt = tracepointStopNames.insert({});
    }
}

return *cacheIt;
}
```

**Review by milianw — CHANGES_REQUESTED** [review]

**Review by milianw — COMMENTED** [review]
I would really like to see some test coverage on this too. ideally again with a perparser-processed data stream as input, exported from a file which actually had a combination of tracepoints:

- sched_switch
- syscall_enter/exit (to check if we match them properly)
- normal tracepoint, without time measurement

File: src/models/data.h [hunk 18]
```diff
@@ -153,7 +161,8 @@ class Costs
     {
         Unknown,
         Tracepoint,
-        Time
+        Time,
+        TracepointCost
```
[the remark this answers is not in the document]
Comment (milianw) [reply src/models/data.h:220]: also:

```
/home/milian/projects/kdab/hotspot/src/flamegraph.cpp: In member function ‘QString FrameGraphicsItem::description() const’:
/home/milian/projects/kdab/hotspot/src/flamegraph.cpp:224:12: warning: enumeration value ‘TracepointCost’ not handled in switch [-Wswitch]
  224 |     switch (root->unit()) {
      |            ^
[96/106] Building CXX object src/CMakeFiles/hotspot.dir/resultsutil.cpp.o
/home/milian/projects/kdab/hotspot/src/resultsutil.cpp: In function ‘void ResultsUtil::hideTracepointColumns(const Data::Costs&, QTreeView*, int)’:
/home/milian/projects/kdab/hotspot/src/resultsutil.cpp:160:16: warning: enumeration value ‘TracepointCost’ not handled in switch [-Wswitch]
  160 |         switch (unit) {
      |                ^
```
